### PR TITLE
Implemented chain tipset and block iterators.

### PR DESCRIPTION
**Pull request by anorth**
My primary goal here is to factor chain traversal out of the MessageQueue so I can re-use
logic for collecting new/old messages in the case of a re-org and expire messages in the
outbound queue. I'm informed the iterators may be useful elsewhere too.

### CODE REVIEW

File: chain/traversal.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+)
+
+// BlockProvider provides blocks. This is a subset of the ReadStore interface.
+type BlockProvider interface {
+	GetBlock(ctx context.Context, cid cid.Cid) (*types.Block, error)
+}
+
+// GetParentTipSet returns the parent tipset of a tipset.
+// The result is empty if the tipset has no parents (including if it is empty itself)
+func GetParentTipSet(ctx context.Context, store BlockProvider, ts types.TipSet) (types.TipSet, error) {
```
Comment (anorth) [inline chain/traversal.go]: This is moved from `message_queue.go`, the only change being the store parameter changing to this narrow interface in place of the concrete `hamt.CborIpldStore`.

File: wallet/wallet_test.go [hunk 2]
```diff
@@ -12,7 +12,7 @@ import (
 	"github.com/filecoin-project/go-filecoin/proofs"
 	"github.com/filecoin-project/go-filecoin/types"
 	"github.com/filecoin-project/go-filecoin/wallet"
-	"github.com/stretchr/testify/require"
+	"gx/ipfs/QmPVkJMTeRC6iBByPWdrRkD3BE5UXsj5HPzb4kPqL186mS/testify/require"
```
Comment (laser) [inline wallet/wallet_test.go]: NON-BLOCKING: Was this change intentional?

File: chain/traversal.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+// the genesis.
+// Blocks from each tipset are yielded in an unspecified but stable order.
+// NOTE: it would be great to align this order with the canonical ordering and define it as such.
+func IterAncestorBlocks(ctx context.Context, store BlockProvider, start types.TipSet) *BlockIterator {
+	tsIter := start.ToSortedCidSet().Iter()
+	return &BlockIterator{TipsetIterator{ctx, store, start, nil}, tsIter}
+}
+
+// BlockIterator is an iterator over blocks in tipsets.
+type BlockIterator struct {
```
Comment (laser) [inline chain/traversal.go]: NON-BLOCKING: Cool idea!

File: chain/traversal_test.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+		b.Height = parents[0].Height + 1
+		b.StateRoot = parents[0].StateRoot
+		for _, p := range parents {
+			b.Parents.Add(p.Cid())
+		}
+	}
+
+	bs.blocks[b.Cid()] = b
+	return b
+
```
Comment (laser) [inline chain/traversal_test.go]: NON-BLOCKING: Extra newline most likely unintentional.

File: core/message_queue.go [hunk 5]
```diff
@@ -5,7 +5,7 @@ import (
 
 	"github.com/filecoin-project/go-filecoin/address"
 	"github.com/filecoin-project/go-filecoin/types"
-	"github.com/pkg/errors"
+	"gx/ipfs/QmVmDhyTTUcQXFD1rRQ64fGLMSAoaQvNH3hwuaCFAPq2hy/errors"
```
Comment (laser) [inline core/message_queue.go]: NON-BLOCKING: Was this change intentional?

File: wallet/wallet_test.go (hunk 2 above)
Comment (phritz) [reply wallet/wallet_test.go]: looks like an improvement to me

File: chain/traversal_test.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
+package chain_test
+
+import (
+	"context"
+	"testing"
+
+	"gx/ipfs/QmPVkJMTeRC6iBByPWdrRkD3BE5UXsj5HPzb4kPqL186mS/testify/assert"
+	"gx/ipfs/QmR8BauakNcBa3RbE4nbQu76PDiJgoQgz8AJdhJuiU4TAw/go-cid"
```
Comment (phritz) [inline chain/traversal_test.go]: imports here and elsewhere are off.

File: chain/traversal_test.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+		ts := requireTipset(t, root)
+		parent, e := chain.GetParentTipSet(ctx, store, ts)
+		require.NoError(e)
+		assert.Empty(parent)
+	})
+	t.Run("plural tipset", func(t *testing.T) {
+		ts := requireTipset(t, b11, b12)
+		parent, e := chain.GetParentTipSet(ctx, store, ts)
+		require.NoError(e)
+		assert.Equal(requireTipset(t, root), parent)
```
Comment (phritz) [inline chain/traversal_test.go]: there is a tipset.equals which is presumably better than default object comparison(?)

File: chain/traversal.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	return it.TipsetIterator.Complete() && it.cidIterator.Complete()
+}
+
+// Next advances the iterator to the next value.
+func (it *BlockIterator) Next() error {
+	it.cidIterator.Next()
+	if it.cidIterator.Complete() {
+		err := it.TipsetIterator.Next()
+		if err == nil {
+			it.cidIterator = it.TipsetIterator.Value().ToSortedCidSet().Iter()
```
Comment (ZenGround0) [inline chain/traversal.go]: It is unclear why blocks are not returned in the canonical order if we are iterating over the tipset in sorted cid order.

File: core/message_pool.go [hunk 9]
```diff
@@ -109,32 +109,12 @@ func NewMessagePool(timer BlockTimer) *MessagePool {
 	}
 }
 
-// getParentTips returns the parent tipset of the provided tipset
-// TODO msgPool should have access to a chain store that can just look this up...
-func getParentTipSet(ctx context.Context, store *hamt.CborIpldStore, ts types.TipSet) (types.TipSet, error) {
```
Comment (ZenGround0) [inline core/message_pool.go]: 🎉

**Comment by ZenGround0** [comment]
At the very least I want to see [walk chain](https://github.com/filecoin-project/go-filecoin/blob/master/chain/default_store.go#L430) (the shared traversal function for BlockHistory and Load) rewritten using this someday soon.  It probably makes sense to replace some or all calls to BlockHistory with this too.  Good deal.

In the near future I want to understand and talk about the tradeoffs between putting these GetIterator functions on the chain package vs. on some object, either the store, the tipset, or some other thing.

File: chain/traversal.go (hunk 8 above)
Comment (anorth) [reply chain/traversal.go]: I couldn't find a bit of code that clearly identified this as the canonical order. Is this the canonical order?

Comment (ZenGround0) [reply chain/traversal.go]: This is a good point.  In retrospect we aren't being good about enforcing one canonical order.  Tipsets are printed and serialized to slices via `ToSlice` in this order but more fundamentally their messages are serialized in the order of increasing ticket bytes.  I think it probably makes sense to enforce the ticket ordering everywhere.

Comment (anorth) [reply chain/traversal.go]: Agreed. I'll leave this and the comment about it here for now.

File: chain/traversal.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+		}
```
[the remark this answers is not in the document]
Comment (anorth) [reply chain/traversal.go]: Thanks, fixed.

**Comment by anorth** [comment]
@ZenGround0 it could go on the store. A free utility function is at least not a wrong place to put it, and easy to move later. It depends on our philosophy toward minimal interfaces.

@acruikshank fair point about the BlockIterator. I wrote it because UpdateMessagePool iterates messages in blocks, and it's easy to stop at a tipset boundary by checking block height. However, I ended up having to iterate tipsets directly. I will retract and stash the BlockIterator for potential use some day.
